### PR TITLE
Fixed --skip-nonexistent option

### LINKED ISSUE #988: Derivation even if missing data
**Describe the bug**
When running the `recipe_preprocessor_derive_test.yml` using the `skip_nonexistent` option, the ESMValCore tries to do the derivation on 0-len cube that correspond to missing files and this results in an error:
`ERROR   esmvalcore.preprocessor:274 Failed to run concatenate([], {})
`

On Mistral, the mising data being:
```
      ctotal:
        <<: *alb_cmip5
        mip: Lmon
        additional_datasets:
          - {dataset: CanESM2, ensemble: r1i1p1}
```

**Please attach**
  - The recipe that you are trying to run, you can find a copy in the `run` directory in the output directory
  - The `main_log_debug.txt` file, this can also be found in the `run` directory in the output directory
[recipe_preprocessor_derive_test.yml.txt](https://github.com/ESMValGroup/ESMValCore/files/5943473/recipe_preprocessor_derive_test.yml.txt)
[main_log_debug.txt](https://github.com/ESMValGroup/ESMValCore/files/5943472/main_log_debug.txt)

### LINKED ISSUE #1040: Concatenate fails when using --skip_nonexistent
```
Traceback (most recent call last):
  File "/home/users/valeriu/anaconda3/envs/release220/lib/python3.9/multiprocessing/pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
  File "/home/users/valeriu/esmvalcore/esmvalcore/_task.py", line 752, in _run_task
    output_files = task.run()
  File "/home/users/valeriu/esmvalcore/esmvalcore/_task.py", line 250, in run
    self.output_files = self._run(input_files)
  File "/home/users/valeriu/esmvalcore/esmvalcore/preprocessor/__init__.py", line 473, in _run
    product.apply(step, self.debug)
  File "/home/users/valeriu/esmvalcore/esmvalcore/preprocessor/__init__.py", line 342, in apply
    self.cubes = preprocess(self.cubes, step, **self.settings[step])
  File "/home/users/valeriu/esmvalcore/esmvalcore/preprocessor/__init__.py", line 286, in preprocess
    result.append(_run_preproc_function(function, items, settings))
  File "/home/users/valeriu/esmvalcore/esmvalcore/preprocessor/__init__.py", line 272, in _run_preproc_function
    return function(items, **kwargs)
  File "/home/users/valeriu/esmvalcore/esmvalcore/preprocessor/_io.py", line 195, in concatenate
    _fix_aux_factories(result)
UnboundLocalError: local variable 'result' referenced before assignment
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/users/valeriu/esmvalcore/esmvalcore/_main.py", line 433, in run
    fire.Fire(ESMValTool())
  File "/home/users/valeriu/anaconda3/envs/release220/lib/python3.9/site-packages/fire/core.py", line 141, in Fire
    component_trace = _Fire(component, args, parsed_flag_args, context, name)
  File "/home/users/valeriu/anaconda3/envs/release220/lib/python3.9/site-packages/fire/core.py", line 466, in _Fire
    component, remaining_args = _CallAndUpdateTrace(
  File "/home/users/valeriu/anaconda3/envs/release220/lib/python3.9/site-packages/fire/core.py", line 681, in _CallAndUpdateTrace
    component = fn(*varargs, **kwargs)
  File "/home/users/valeriu/esmvalcore/esmvalcore/_main.py", line 410, in run
    process_recipe(recipe_file=recipe, config_user=cfg)
  File "/home/users/valeriu/esmvalcore/esmvalcore/_main.py", line 104, in process_recipe
    recipe.run()
  File "/home/users/valeriu/esmvalcore/esmvalcore/_recipe.py", line 1366, in run
    self.tasks.run(max_parallel_tasks=self._cfg['max_parallel_tasks'])
  File "/home/users/valeriu/esmvalcore/esmvalcore/_task.py", line 674, in run
    self._run_parallel(max_parallel_tasks)
  File "/home/users/valeriu/esmvalcore/esmvalcore/_task.py", line 717, in _run_parallel
    _copy_results(task, running[task])
  File "/home/users/valeriu/esmvalcore/esmvalcore/_task.py", line 740, in _copy_results
    task.output_files, updated_products = future.get()
  File "/home/users/valeriu/anaconda3/envs/release220/lib/python3.9/multiprocessing/pool.py", line 771, in get
    raise self._value
  File "/home/users/valeriu/anaconda3/envs/release220/lib/python3.9/multiprocessing/pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
  File "/home/users/valeriu/esmvalcore/esmvalcore/_task.py", line 752, in _run_task
    output_files = task.run()
  File "/home/users/valeriu/esmvalcore/esmvalcore/_task.py", line 250, in run
    self.output_files = self._run(input_files)
  File "/home/users/valeriu/esmvalcore/esmvalcore/preprocessor/__init__.py", line 473, in _run
    product.apply(step, self.debug)
  File "/home/users/valeriu/esmvalcore/esmvalcore/preprocessor/__init__.py", line 342, in apply
    self.cubes = preprocess(self.cubes, step, **self.settings[step])
  File "/home/users/valeriu/esmvalcore/esmvalcore/preprocessor/__init__.py", line 286, in preprocess
    result.append(_run_preproc_function(function, items, settings))
  File "/home/users/valeriu/esmvalcore/esmvalcore/preprocessor/__init__.py", line 272, in _run_preproc_function
    return function(items, **kwargs)
  File "/home/users/valeriu/esmvalcore/esmvalcore/preprocessor/_io.py", line 195, in concatenate
    _fix_aux_factories(result)
UnboundLocalError: local variable 'result' referenced before assignment
2021-03-10 12:56:37,266 UTC [27481] INFO    
If you have a question or need help, please start a new discussion on https://github.com/ESMValGroup/ESMValTool/discussions
If you suspect this is a bug, please open an issue on https://github.com/ESMValGroup/ESMValTool/issues
To make it easier to find out what the problem is, please consider attaching the files run/recipe_*.yml and run/main_log_debug.txt from the output directory.
```
While it's trying to concatenate an empty list of cubes because there is no data - have we lost the interface between IO and recipe? Looking into this now...

### PR DESCRIPTION
**Pull request by schlunma**
This PR fixes a bug that was introduced in #917 that prevented the usage of the ``--skip-nonexistent`` option. I also add a tiny test for it.

Note that I also adapted the preprocessors `concatenate` and `save` which gave ugly error messages when an empty cube was passed. This should never happen during a regular ESMValTool run, but could happen when these functions are used as part of the `esmvalcore` API. 

Closes #1040, closes #988.

Link to documentation:

***

## [Before you get started](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#getting-started)

- [x] [☝ Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do

## [Checklist](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#checklist-for-pull-requests)

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [x] [🧪][2] The new functionality is [relevant and scientifically sound](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#scientific-relevance)
- [x] [🛠][1] This pull request has a [descriptive title and labels](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-title-and-label)
- [x] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#code-quality)
- [x] [🧪][2] and [🛠][1] [Documentation](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#documentation) is available
- [x] [🛠][1] [Unit tests](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#tests) have been added
- [x] [🛠][1] Changes are [backward compatible](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#backward-compatibility)
- [x] [🛠][1] Any changed [dependencies have been added or removed](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#dependencies) correctly
- [x] [🛠][1] The [list of authors](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#list-of-authors) is up to date
- [x] [🛠][1] All [checks below this pull request](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-checks) were successful

***

To help with the number pull requests:

- 🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValCore/pulls) in this repository

### CODE REVIEW

**Comment by bouweandela** [comment]
@jvegasbsc Would you have time to review this?

File: esmvalcore/preprocessor/_io.py [hunk 1]
```diff
@@ -173,6 +173,8 @@ def _get_concatenation_error(cubes):
 
 def concatenate(cubes):
     """Concatenate all cubes after fixing metadata."""
+    if not cubes:
+        return cubes
```
Comment (valeriupredoi) [inline esmvalcore/preprocessor/_io.py:177]: thank you! I meant to plug this in since January :laughing:

**Review by valeriupredoi — APPROVED** [review]
very nice, thanks a lot @schlunma :beer:

**Comment by valeriupredoi** [comment]
this is good, @jvegasbsc have a look too pls mate but it's good to go by me :+1:

**Pull request merged**